File: 5-Problem-A-Day/17. Letter Combinations of a Phone Number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<>();
        List<String> curr = new ArrayList<>();
        String[][] nums = {{}, {}, {"a", "b", "c"}, {"d", "e", "f"}, {"g", "h", "i"},
                {"j", "k", "l"}, {"m", "n", "o"}, {"p", "q", "r", "s"}, {"t", "u", "v"},
                {"w", "x", "y", "z"}};
        for (int i = 0; i < digits.length(); i++) {
            int d = Character.getNumericValue(digits.charAt(i));
            if (ans.size() == 0) {
                for (String n : nums[d])
                    ans.add(n);
            } else {
                for (String a : ans) {
                    for (String n : nums[d])
                        curr.add(a + n);
                }
                ans = new ArrayList<>(curr);
                curr = new ArrayList<>();
            }
        }
        return ans;
    }
}

File: 2024/17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    unordered_map<char, vector<char>> cm = {
        {'2', {'a', 'b', 'c'}}, {'3', {'d', 'e', 'f'}},
        {'4', {'g', 'h', 'i'}}, {'5', {'j', 'k', 'l'}},
        {'6', {'m', 'n', 'o'}}, {'7', {'p', 'q', 'r', 's'}},
        {'8', {'t', 'u', 'v'}}, {'9', {'w', 'x', 'y', 'z'}}};
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) {
            return {};
        }
        vector<string> ret = {""};
        for (char d : digits) {
            vector<string> curr;
            for (char c : cm[d]) {
                for (string& cu : ret) {
                    curr.push_back(cu + c);
                }
            }
            ret = curr;
        }
        return ret;
    }
};